source/common/event/scaled_range_timer.cc

```diff
@@ -49,12 +49,21 @@ class ScaledRangeTimerManager::RangeTimerImpl final : public RangeTimer {
     scope_ = scope;
     ENVOY_LOG_MISC(trace, "enableTimer called on {} for ({}ms, {}ms)", static_cast<void*>(this),
                    min_ms.count(), max_ms.count());
-    if (min_ms <= std::chrono::milliseconds::zero()) {
-      auto it = manager_.add(*this, max_ms);
-      state_.emplace<ScalingMax>(max_ms, it);
+
+    // Align the actual required wait time with a bucket boundary.
+    BucketHandle bucket_handle =
+        manager_.getBucketForDuration(max_ms - std::max(std::chrono::milliseconds::zero(), min_ms));
+    const std::chrono::milliseconds scaling_duration = manager_.getBucketDuration(bucket_handle);
+    const std::chrono::milliseconds wait_time = max_ms - scaling_duration;
+
+    // Now wait_time + scaling_duration = max_ms, and scaling_duration is aligned on the boundary of
+    // bucket `bucket_handle`.
+    if (wait_time <= std::chrono::milliseconds::zero()) {
+      auto it = manager_.add(*this, bucket_handle);
+      state_.emplace<ScalingMax>(bucket_handle, it);
```

Comment:
I encourage you to seek feedback from others on this comment thread: https://github.com/envoyproxy/envoy/pull/12414#discussion_r465758394